src/librustc/dep_graph/dep_node.rs

```diff
@@ -127,6 +127,11 @@ pub enum DepNode<D: Clone + Debug> {
     TraitItems(D),
     ReprHints(D),
     TraitSelect(Vec<D>),
+
+    // An optional alternative to `TraitSelect` that avoids a heap allocation
+    // in the case where there is a single D. (Note that `TraitSelect` is still
+    // allowed to contain a Vec with a single D.)
```

Comment:
I think I tried them and had problems, but now I'm not 100% sure why. I guess because you can't "map" them as easily. You'd have to refactor, and I'm not sure I see the point. The main thing is to ensure that `Vec: Clone` doesn't interfere with `String: Clone` and the like.